File: app/src/main/java/fr/free/nrw/commons/description/DescriptionEditActivity.java
package fr.free.nrw.commons.description;

import static fr.free.nrw.commons.description.EditDescriptionConstants.LIST_OF_DESCRIPTION_AND_CAPTION;
import static fr.free.nrw.commons.description.EditDescriptionConstants.UPDATED_WIKITEXT;
import static fr.free.nrw.commons.description.EditDescriptionConstants.WIKITEXT;

import android.app.ProgressDialog;
import android.content.Intent;
import android.os.Bundle;
import android.os.Parcelable;
import android.widget.ImageView;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.AppCompatButton;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Optional;
import fr.free.nrw.commons.R;
import fr.free.nrw.commons.kvstore.JsonKvStore;
import fr.free.nrw.commons.upload.UploadMediaDetail;
import fr.free.nrw.commons.upload.UploadMediaDetailAdapter;
import fr.free.nrw.commons.utils.DialogUtil;
import java.util.ArrayList;
import java.util.List;
import javax.inject.Inject;
import javax.inject.Named;

/**
 * Activity for populating and editing existing description and caption
 */
public class DescriptionEditActivity extends AppCompatActivity implements
    UploadMediaDetailAdapter.EventListener {

    /**
     * Adapter for showing UploadMediaDetail in the activity
     */
    private UploadMediaDetailAdapter uploadMediaDetailAdapter;

    /**
     * For getting default preference
     */
    @Inject
    @Named("default_preferences")
    JsonKvStore defaultKvStore;

    /**
     * Recyclerview for recycling data in views
     */
    @BindView(R.id.rv_descriptions_captions)
    RecyclerView rvDescriptions;

    /**
     * Submit button for submitting new wikiText
     */
    @BindView(R.id.btn_edit_submit)
    AppCompatButton btnSubmit;

//    @BindView(R.id.mapbox_place_picker_toolbar_back_button)
//    ImageView backButton;


    /**
     * Current wikitext
     */
    String wikiText;
    ProgressDialog progressDialog;

    @Override
    protected void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_description_edit);

        ButterKnife.bind(this, this);
        final Bundle bundle = getIntent().getExtras();
        final ArrayList<UploadMediaDetail> descriptionAndCaptions
            = bundle.getParcelableArrayList(LIST_OF_DESCRIPTION_AND_CAPTION);
        wikiText = bundle.getString(WIKITEXT);

        initRecyclerView(descriptionAndCaptions);
    }

    /**
     * Initializes the RecyclerView
     * @param descriptionAndCaptions list of description and caption
     */
    private void initRecyclerView(final ArrayList<UploadMediaDetail> descriptionAndCaptions) {
        uploadMediaDetailAdapter
            = new UploadMediaDetailAdapter("en",descriptionAndCaptions);
        uploadMediaDetailAdapter.setCallback(this::showInfoAlert);
        uploadMediaDetailAdapter.setEventListener(this);
        rvDescriptions.setLayoutManager(new LinearLayoutManager(this));
        rvDescriptions.setAdapter(uploadMediaDetailAdapter);
    }

    /**
     * show dialog with info
     * @param titleStringID Title ID
     * @param messageStringId Message ID
     */
    private void showInfoAlert(final int titleStringID, final int messageStringId) {
        DialogUtil.showAlertDialog(this, getString(titleStringID),
            getString(messageStringId), getString(android.R.string.ok),
            null, true);
    }

    @Override
    public void onPrimaryCaptionTextChange(final boolean isNotEmpty) {

    }

    @Optional
    @OnClick(R.id.mapbox_place_picker_toolbar_back_button)
    public void onBackButtonClicked(){
        onBackPressed();
    }

    @OnClick(R.id.btn_add_description)
    public void onButtonAddDescriptionClicked() {
        final UploadMediaDetail uploadMediaDetail = new UploadMediaDetail();
        uploadMediaDetail.setManuallyAdded(true);//This was manually added by the user
        uploadMediaDetailAdapter.addDescription(uploadMediaDetail);
        rvDescriptions.smoothScrollToPosition(uploadMediaDetailAdapter.getItemCount()-1);
    }


    @OnClick(R.id.btn_edit_submit)
    public void onSubmitButtonClicked(){
        showLoggingProgressBar();
        final List<UploadMediaDetail> uploadMediaDetails = uploadMediaDetailAdapter.getItems();
        updateDescription(uploadMediaDetails);
        finish();
    }

    /**
     * Updates newly added descriptions in the wikiText and send to calling fragment
     * @param uploadMediaDetails descriptions and captions
     */
    private void updateDescription(final List<UploadMediaDetail> uploadMediaDetails) {
        int descriptionIndex = wikiText.indexOf("description=");
        if (descriptionIndex == -1){
            descriptionIndex = wikiText.indexOf("Description=");
        }

        final StringBuilder buffer = new StringBuilder();

        if (descriptionIndex != -1) {

            final String descriptionStart = wikiText.substring(0, descriptionIndex + 12);
            final String descriptionToEnd = wikiText.substring(descriptionIndex+12);
            final int descriptionEndIndex = descriptionToEnd.indexOf("\n");
            final String descriptionEnd = wikiText.substring(descriptionStart.length()
                +descriptionEndIndex);

            buffer.append(descriptionStart);
            for (int i=0; i<uploadMediaDetails.size(); i++) {
                final UploadMediaDetail uploadDetails = uploadMediaDetails.get(i);
                if (!uploadDetails.getDescriptionText().equals("")) {
                    buffer.append("{{");
                    buffer.append(uploadDetails.getLanguageCode());
                    buffer.append("|1=");
                    buffer.append(uploadDetails.getDescriptionText());
                    buffer.append("}}, ");
                }
            }
            buffer.deleteCharAt(buffer.length()-1);
            buffer.deleteCharAt(buffer.length()-1);
            buffer.append(descriptionEnd);
        }
        final Intent returningIntent = new Intent();
        returningIntent.putExtra(UPDATED_WIKITEXT, buffer.toString());
        returningIntent.putParcelableArrayListExtra(LIST_OF_DESCRIPTION_AND_CAPTION,
            (ArrayList<? extends Parcelable>) uploadMediaDetails);
        setResult(RESULT_OK, returningIntent);
        finish();
    }

    private void showLoggingProgressBar() {
        progressDialog = new ProgressDialog(this);
        progressDialog.setIndeterminate(true);
        progressDialog.setTitle(getString(R.string.updating_caption_title));
        progressDialog.setMessage(getString(R.string.updating_caption_message));
        progressDialog.setCanceledOnTouchOutside(false);
        progressDialog.show();
    }
}

File: app/src/main/java/fr/free/nrw/commons/description/DescriptionEditActivity.kt
package fr.free.nrw.commons.description

import android.app.ProgressDialog
import android.content.Intent
import android.os.Bundle
import android.os.Parcelable
import android.speech.RecognizerIntent
import android.view.View
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import fr.free.nrw.commons.Media
import fr.free.nrw.commons.R
import fr.free.nrw.commons.databinding.ActivityDescriptionEditBinding
import fr.free.nrw.commons.description.EditDescriptionConstants.LIST_OF_DESCRIPTION_AND_CAPTION
import fr.free.nrw.commons.description.EditDescriptionConstants.WIKITEXT
import fr.free.nrw.commons.recentlanguages.RecentLanguagesDao
import fr.free.nrw.commons.settings.Prefs
import fr.free.nrw.commons.theme.BaseActivity
import fr.free.nrw.commons.upload.UploadMediaDetail
import fr.free.nrw.commons.upload.UploadMediaDetailAdapter
import fr.free.nrw.commons.utils.DialogUtil.showAlertDialog
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.functions.Consumer
import io.reactivex.schedulers.Schedulers
import timber.log.Timber
import javax.inject.Inject

/**
 * Activity for populating and editing existing description and caption
 */
class DescriptionEditActivity : BaseActivity(), UploadMediaDetailAdapter.EventListener {
    /**
     * Adapter for showing UploadMediaDetail in the activity
     */
    private lateinit var uploadMediaDetailAdapter: UploadMediaDetailAdapter

    /**
     * Recyclerview for recycling data in views
     */
    @JvmField
    var rvDescriptions: RecyclerView? = null

    /**
     * Current wikitext
     */
    var wikiText: String? = null

    /**
     * Media object
     */
    var media: Media? = null

    /**
     * Saved language
     */
    private lateinit var savedLanguageValue: String

    /**
     * For showing progress dialog
     */
    private var progressDialog: ProgressDialog? = null

    @Inject
    lateinit var recentLanguagesDao: RecentLanguagesDao

    private lateinit var binding: ActivityDescriptionEditBinding

    private val REQUEST_CODE_FOR_VOICE_INPUT = 1213

    private var descriptionAndCaptions: ArrayList<UploadMediaDetail>? = null

    @Inject lateinit var descriptionEditHelper: DescriptionEditHelper

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityDescriptionEditBinding.inflate(layoutInflater)
        setContentView(binding.root)

        val bundle = intent.extras

        if (savedInstanceState != null) {
            descriptionAndCaptions = savedInstanceState.getParcelableArrayList(LIST_OF_DESCRIPTION_AND_CAPTION)
            wikiText = savedInstanceState.getString(WIKITEXT)
            savedLanguageValue = savedInstanceState.getString(Prefs.DESCRIPTION_LANGUAGE)!!
            media = savedInstanceState.getParcelable("media")
        } else {
            descriptionAndCaptions =
                bundle!!.getParcelableArrayList(LIST_OF_DESCRIPTION_AND_CAPTION)!!
            wikiText = bundle.getString(WIKITEXT)
            savedLanguageValue = bundle.getString(Prefs.DESCRIPTION_LANGUAGE)!!
            media = bundle.getParcelable("media")
        }
        initRecyclerView(descriptionAndCaptions)

        binding.btnEditSubmit.setOnClickListener(::onSubmitButtonClicked)
        binding.toolbarBackButton.setOnClickListener(::onBackButtonClicked)
    }

    /**
     * Initializes the RecyclerView
     * @param descriptionAndCaptions list of description and caption
     */
    private fun initRecyclerView(descriptionAndCaptions: ArrayList<UploadMediaDetail>?) {
        uploadMediaDetailAdapter = UploadMediaDetailAdapter(this,
            savedLanguageValue, descriptionAndCaptions, recentLanguagesDao)
        uploadMediaDetailAdapter.setCallback { titleStringID: Int, messageStringId: Int ->
            showInfoAlert(
                titleStringID,
                messageStringId
            )
        }
        uploadMediaDetailAdapter.setEventListener(this)
        rvDescriptions = binding.rvDescriptionsCaptions
        rvDescriptions!!.layoutManager = LinearLayoutManager(this)
        rvDescriptions!!.adapter = uploadMediaDetailAdapter
    }

    /**
     * show dialog with info
     * @param titleStringID Title ID
     * @param messageStringId Message ID
     */
    private fun showInfoAlert(titleStringID: Int, messageStringId: Int) {
        showAlertDialog(
            this, getString(titleStringID),
            getString(messageStringId), getString(android.R.string.ok),
            null, true
        )
    }

    override fun onPrimaryCaptionTextChange(isNotEmpty: Boolean) {}

    /**
     * Adds new language item to RecyclerView
     */
    override fun addLanguage() {
        val uploadMediaDetail = UploadMediaDetail()
        uploadMediaDetail.isManuallyAdded = true //This was manually added by the user
        uploadMediaDetailAdapter.addDescription(uploadMediaDetail)
        rvDescriptions!!.smoothScrollToPosition(uploadMediaDetailAdapter.itemCount - 1)
    }

    private fun onBackButtonClicked(view: View) {
       onBackPressedDispatcher.onBackPressed()
    }

    private fun onSubmitButtonClicked(view: View) {
        showLoggingProgressBar()
        val uploadMediaDetails = uploadMediaDetailAdapter.items
        updateDescription(uploadMediaDetails)
        finish()
    }

    /**
     * Updates newly added descriptions in the wikiText and send to calling fragment
     * @param uploadMediaDetails descriptions and captions
     */
    private fun updateDescription(uploadMediaDetails: List<UploadMediaDetail?>) {
        var descriptionIndex = wikiText!!.indexOf("description=")
        if (descriptionIndex == -1) {
            descriptionIndex = wikiText!!.indexOf("Description=")
        }
        val buffer = StringBuilder()
        if (descriptionIndex != -1) {
            val descriptionStart = wikiText!!.substring(0, descriptionIndex + 12)
            val descriptionToEnd = wikiText!!.substring(descriptionIndex + 12)
            val descriptionEndIndex = descriptionToEnd.indexOf("\n")
            val descriptionEnd = wikiText!!.substring(
                descriptionStart.length
                        + descriptionEndIndex
            )
            buffer.append(descriptionStart)
            for (i in uploadMediaDetails.indices) {
                val uploadDetails = uploadMediaDetails[i]
                if (uploadDetails!!.descriptionText != "") {
                    buffer.append("{{")
                    buffer.append(uploadDetails.languageCode)
                    buffer.append("|1=")
                    buffer.append(uploadDetails.descriptionText)
                    buffer.append("}}")
                }
            }
            buffer.replace(", $".toRegex(), "")
            buffer.append(descriptionEnd)
        }
        editDescription(media!!, buffer.toString(), uploadMediaDetails as ArrayList<UploadMediaDetail>)

        finish()
    }

    /**
     * Edits description and caption
     * @param media media object
     * @param updatedWikiText updated wiki text
     * @param uploadMediaDetails descriptions and captions
     */
    private fun editDescription(media : Media, updatedWikiText : String, uploadMediaDetails : ArrayList<UploadMediaDetail>){
        descriptionEditHelper?.addDescription(
            applicationContext, media,
            updatedWikiText
        )
            ?.subscribeOn(Schedulers.io())
            ?.observeOn(AndroidSchedulers.mainThread())
            ?.subscribe(Consumer<Boolean> { s: Boolean? -> Timber.d("Descriptions are added.") })?.let {
                compositeDisposable.add(
                    it
            )
            }

        val updatedCaptions = LinkedHashMap<String, String>()
        for (mediaDetail in uploadMediaDetails) {
            compositeDisposable.add(
                descriptionEditHelper!!.addCaption(
                    applicationContext, media,
                    mediaDetail.languageCode, mediaDetail.captionText
                )
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe { s: Boolean? ->
                        updatedCaptions[mediaDetail.languageCode!!] = mediaDetail.captionText
                        media.captions = updatedCaptions
                        Timber.d("Caption is added.")
                    })
        }
    }

    private fun showLoggingProgressBar() {
        progressDialog = ProgressDialog(this)
        progressDialog!!.isIndeterminate = true
        progressDialog!!.setTitle(getString(R.string.updating_caption_title))
        progressDialog!!.setMessage(getString(R.string.updating_caption_message))
        progressDialog!!.setCanceledOnTouchOutside(false)
        progressDialog!!.show()
    }

    override
    fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == REQUEST_CODE_FOR_VOICE_INPUT) {
            if (resultCode == RESULT_OK && data != null) {
                val result = data.getStringArrayListExtra( RecognizerIntent.EXTRA_RESULTS )
                uploadMediaDetailAdapter.handleSpeechResult(result!![0]) }
            else { Timber.e("Error %s", resultCode) }
        }
    }
    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)

        outState.putParcelableArrayList(LIST_OF_DESCRIPTION_AND_CAPTION, uploadMediaDetailAdapter.items as ArrayList<out Parcelable?>)
        outState.putString(WIKITEXT, wikiText)
        outState.putString(Prefs.DESCRIPTION_LANGUAGE, savedLanguageValue)
        //save Media
        outState.putParcelable("media", media)
    }
}
